src/Microsoft.AspNetCore.SignalR.Client/RpcConnection.cs

```diff
@@ -66,57 +90,86 @@ public void Dispose()
             _connection.Dispose();
         }
 
-        // TODO: Clean up the API here. Negotiation of format would be better than providing an adapter instance.
-        public static Task<RpcConnection> ConnectAsync(Uri url, IInvocationAdapter adapter, ITransport transport, PipelineFactory pipelineFactory) => ConnectAsync(url, transport, new HttpClient(), pipelineFactory, NullLoggerFactory.Instance);
-        public static Task<RpcConnection> ConnectAsync(Uri url, IInvocationAdapter adapter, ITransport transport, PipelineFactory pipelineFactory, ILoggerFactory loggerFactory) => ConnectAsync(url, transport, new HttpClient(), pipelineFactory, loggerFactory);
-        public static Task<RpcConnection> ConnectAsync(Uri url, IInvocationAdapter adapter, ITransport transport, HttpClient httpClient, PipelineFactory pipelineFactory) => ConnectAsync(url, transport, httpClient, pipelineFactory, NullLoggerFactory.Instance);
+        // TODO: Clean up the API here. Negotiation of format would be better than providing an adapter instance. Similarly, we should not require a logger factory
+        public static Task<RpcConnection> ConnectAsync(Uri url, IInvocationAdapter adapter, ITransport transport, PipelineFactory pipelineFactory, ILoggerFactory loggerFactory) => ConnectAsync(url, adapter, transport, new HttpClient(), pipelineFactory, loggerFactory);
 
         public static async Task<RpcConnection> ConnectAsync(Uri url, IInvocationAdapter adapter, ITransport transport, HttpClient httpClient, PipelineFactory pipelineFactory, ILoggerFactory loggerFactory)
         {
             // Connect the underlying connection
             var connection = await Connection.ConnectAsync(url, transport, httpClient, pipelineFactory, loggerFactory);
 
             // Create the RPC connection wrapper
-            return new RpcConnection(connection, adapter);
+            return new RpcConnection(connection, adapter, loggerFactory.CreateLogger<RpcConnection>());
         }
 
         private async Task ReceiveMessages(CancellationToken cancellationToken)
         {
+            await Task.Yield();
+
+            _logger.LogTrace("Beginning receive loop");
             while (!cancellationToken.IsCancellationRequested)
             {
                 // This is a little odd... we want to remove the InvocationRequest once and only once so we pull it out in the callback,
                 // and stash it here because we know the callback will have finished before the end of the await.
-                InvocationRequest irq = default(InvocationRequest);
-                var result = await _adapter.ReadInvocationResultDescriptorAsync(_stream, id =>
+                var message = await _adapter.ReadMessageAsync(_stream, this, cancellationToken);
+
+                var invocationDescriptor = message as InvocationDescriptor;
```

Comment:
(When doing protobuf) I was wondering if we can have some kind of base class that would have the message type inside to avoid this kind of casts.